File: backend/app/models/ml_training.py
"""
ML Training Models
Pydantic models for ML training projects, annotations, and trained models.
"""
from typing import Any, Dict, List, Optional
from datetime import datetime
from pydantic import BaseModel, Field
from bson import ObjectId


class PyObjectId(str):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        if not ObjectId.is_valid(v):
            raise ValueError("Invalid ObjectId")
        return str(v)


# ─────────────────────────────────── Segment / Annotation structures ──

class CharSegment(BaseModel):
    """One auto-segmented character box with an optional OK/NG label."""
    id: str
    x: float   # normalized 0-1, relative to full image
    y: float
    w: float
    h: float
    label: Optional[str] = None  # "OK" | "NG" | None
    char_id: Optional[str] = None  # "A", "B", "0", "/"... — enables per-char golden template


class AnnotationRegion(BaseModel):
    """User-drawn region containing auto-segmented character boxes."""
    id: str
    x: float   # normalized 0-1, relative to full image
    y: float
    w: float
    h: float
    segments: List[CharSegment] = []


# ─────────────────────────────────── ML Project ──────────────────────

class MLProjectCreate(BaseModel):
    name: str
    description: Optional[str] = None


class MLProjectUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


class MLProjectInDB(BaseModel):
    id: str = Field(alias="_id")
    name: str
    description: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    image_count: int = 0
    labeled_count: int = 0   # images that have at least one labeled segment
    status: str = "active"   # active | training | trained

    model_config = {"populate_by_name": True}


# ─────────────────────────────────── Annotation ──────────────────────

class MLAnnotationSave(BaseModel):
    regions: List[AnnotationRegion] = []


class MLAnnotationInDB(BaseModel):
    id: str = Field(alias="_id")
    project_id: str
    filename: str
    regions: List[AnnotationRegion] = []
    updated_at: datetime

    model_config = {"populate_by_name": True}


# ─────────────────────────────────── Trained Model ───────────────────

class MLModelMetrics(BaseModel):
    accuracy_train: float = 0.0
    accuracy_test: float = 0.0
    n_ok: int = 0
    n_ng: int = 0
    n_total: int = 0
    confusion_matrix: List[List[int]] = []
    report: str = ""


class MLModelInDB(BaseModel):
    id: str = Field(alias="_id")
    project_id: str
    algorithm: str          # "rf" | "svm" | "mlp"
    params: Dict[str, Any] = {}
    augment_factor: int = 0  # 0=off, 2=x2 ...
    metrics: MLModelMetrics = MLModelMetrics()
    model_path: str = ""
    status: str = "pending"  # pending | training | completed | failed
    error: Optional[str] = None
    created_at: datetime
    # Live-progress fields populated during training (used by the FE log panel)
    phase: Optional[str] = None        # "preparing" | "embedding" | "training" | "evaluating" | "saving" | "completed"
    progress: float = 0.0              # 0..100

    model_config = {"populate_by_name": True}


# ─────────────────────────────────── Request / Response bodies ───────

class SyntheticOkOptions(BaseModel):
    font_paths: Optional[List[str]] = None
    style_sample_n: int = 64
    sample_strategy: str = "random"
    rotation_max_deg: float = 5.0
    size_jitter: float = 0.30
    char_fill_min: float = 0.85
    char_fill_max: float = 0.95
    bg_per_char: int = 24
    fill_min: float = 0.10
    fill_max: float = 0.65
    min_contrast: float = 20.0
    max_retries: int = 4
    # When True, the project-derived glyph dictionary is added to the font
    # pool alongside any selected TTFs. Glyph dict is loaded from disk —
    # build/rebuild via POST /ml/projects/{id}/glyphs/rebuild.
    use_project_glyphs: bool = False


class TrainRequest(BaseModel):
    algorithm: str = "rf"
    augment_factor: int = 0
    test_split: float = 0.2
    threshold: float = 0.5
    n_estimators: int = 100
    max_iter: int = 500
    C: float = 1.0
    hidden_layer_sizes: List[int] = [128, 64]
    centroid_temperature: float = 5.0
    severity_dist: Optional[Dict[str, float]] = None
    ok_synth_target: int = 0
    include_imported_chars: bool = True
    synth_ok_options: Optional[SyntheticOkOptions] = None
    # Whitelist of NG defect types to apply during augmentation. None = all.
    enabled_defect_types: Optional[List[str]] = None


class SyntheticPreviewRequest(BaseModel):
    augment_factor: int             # must be >= 2
    label: str = "NG"               # Only 'NG' is supported (OK aug removed)
    severity_dist: Optional[Dict[str, float]] = None
    # When set, every generated crop applies exactly this defect type. Used by
    # the NG options modal to preview a single error class at a time.
    force_defect_type: Optional[str] = None
    # Optional char_id whitelist — restricts source OK samples used to render
    # the preview (per-defect previews use a single char to keep grid small).
    char_filter: Optional[List[str]] = None
    # When set, only these defect types are eligible during augmentation
    # (overrides the default 10-type pool). None = all types enabled.
    enabled_defect_types: Optional[List[str]] = None


class TestSetRequest(BaseModel):
    model_id: str


class SegmentRequest(BaseModel):
    filename: str
    region: Dict[str, float]        # {x, y, w, h} normalized


class LabeledCrop(BaseModel):
    segment_id: str
    region_id: str
    filename: str
    label: Optional[str]
    crop_b64: str                   # base64 JPEG
    char_id: Optional[str] = None   # for FE badge + golden template


class PredictResult(BaseModel):
    segment_id: str
    x: float
    y: float
    w: float
    h: float
    prob_ok: float
    label: str                      # "OK" | "NG"
    crop_b64: str
    char_id: Optional[str] = None
    # Diff heatmap preview (only when char_id has a golden in model)
    aligned_b64: Optional[str] = None
    golden_b64: Optional[str] = None
    diff_b64: Optional[str] = None


class CharCoverageResponse(BaseModel):
    covered: List[str]
    missing: List[str]
    coverage_pct: float
    model_chars: List[str]              # full list of chars the model has goldens for


class PredictResponse(BaseModel):
    model_id: str
    algorithm: str
    results: List[PredictResult]


# ─────────────────────────────────── Char Imports ────────────────────
#
# Active-learning pool: chars cropped from past inspection results, stored as
# JPEG files on disk under public/ml_projects/{pid}/imported_chars/{batch}/.
# Train pipeline reads from this pool in addition to Label-tab annotations.

class MLCharImportBatchInDB(BaseModel):
    id: str = Field(alias="_id")
    project_id: str
    name: str                                  # auto: "Import 2026-05-08 14:30" (renameable)
    created_at: datetime

    model_config = {"populate_by_name": True}


class MLCharImportInDB(BaseModel):
    id: str = Field(alias="_id")
    project_id: str
    batch_id: str

    # Provenance — used for dedup against future imports.
    inspection_id: str
    annotation_idx: int
    frame_idx: int
    camera_serial: str
    recipe_id: Optional[str] = None
    recipe_name: Optional[str] = None

    # Editable in the Imported Chars tab.
    char_id: Optional[str] = None
    label: str = "NG"                          # "OK" | "NG"

    # Crop file relative to /public root (served via /api/ml-files static mount).
    crop_path: str

    # Display context — what the model originally said.
    ml_label: str = "NG"                       # "OK" | "NG"
    ml_p_ok: float = 0.0
    source_timestamp: Optional[datetime] = None

    created_at: datetime
    updated_at: datetime

    model_config = {"populate_by_name": True}


class CharImportSelection(BaseModel):
    inspection_id: str
    annotation_idx: int
    # camera_serial + frame_idx scope the selection to ONE specific char crop —
    # a single inspection has multiple frames sharing annotation_idx values.
    camera_serial: str = ""
    frame_idx: int = 0


class CharImportCreateRequest(BaseModel):
    selections: List[CharImportSelection]
    batch_name: Optional[str] = None           # auto-generated if omitted


class CharImportBatchUpdate(BaseModel):
    name: str


class CharImportUpdate(BaseModel):
    char_id: Optional[str] = None
    label: Optional[str] = None                # "OK" | "NG"


class CharImportBulkUpdate(BaseModel):
    char_ids: List[str]                        # doc IDs in ml_char_imports
    label: Optional[str] = None                # set this label on all
    delete: bool = False                       # if True, delete instead


class CharImportBatchResponse(BaseModel):
    id: str
    name: str
    created_at: datetime
    total: int
    ok_count: int
    ng_count: int


class CharImportItemResponse(BaseModel):
    id: str
    batch_id: str
    char_id: Optional[str]
    label: str
    crop_url: str                              # full URL via /api/ml-files static mount
    ml_label: str
    ml_p_ok: float
    inspection_id: str
    annotation_idx: int
    recipe_name: Optional[str]
    camera_serial: str
    frame_idx: int
    source_timestamp: Optional[datetime]
    created_at: datetime

File: frontend-ts/src/services/mlTraining.ts
import api, { API_BASE_URL } from './http';

// Server origin (e.g. "http://localhost:8000") — used to build static image URLs
const SERVER_ORIGIN = API_BASE_URL.replace(/\/api\/?$/, '');

// ──────── Types ────────────────────────────────────────────────────────────

export interface MLProject {
  id: string;
  name: string;
  description?: string;
  created_at: string;
  updated_at: string;
  created_by?: string;
  image_count: number;
  labeled_count: number;
  status: 'active' | 'training' | 'trained';
}

export interface AvailableImage {
  filename: string;
  url: string;
}

export interface ProjectImage {
  filename: string;
  url: string;
  has_annotation: boolean;
}

export interface CharSegment {
  id: string;
  x: number; y: number; w: number; h: number;
  label?: 'OK' | 'NG' | null;
  char_id?: string | null;   // auto-populated from recipe expected_text or manually set
}

export interface AnnotationRegion {
  id: string;
  x: number; y: number; w: number; h: number;
  segments: CharSegment[];
}

export interface AnnotationData {
  project_id: string;
  filename: string;
  regions: AnnotationRegion[];
}

export interface LabeledCrop {
  segment_id: string;
  region_id: string;
  filename: string;
  label: 'OK' | 'NG';
  crop_b64: string;
  char_id?: string | null;
}

export interface SyntheticCrop {
  source_segment_id: string;
  filename: string;
  label: 'OK' | 'NG';
  crop_b64: string;
  char_id?: string | null;
  aug_type?: string;            // realistic NG types or legacy tags
}

export interface SyntheticOkCrop {
  crop_b64: string;
  char_id: string;
  font_name: string;
  rotation_deg: number;
  source: 'synthetic_ok';
}

export type MLAlgorithm = 'rf' | 'svm' | 'mlp' | 'centroid';

export interface SeverityDist {
  subtle: number;
  light: number;
  medium: number;
  heavy: number;
}

export interface TrainRequest {
  algorithm: MLAlgorithm;
  augment_factor: number;
  threshold?: number;         // prob_ok >= threshold → OK
  test_split?: number;
  n_estimators?: number;      // RF only
  max_iter?: number;
  C?: number;
  hidden_layer_sizes?: number[];
  severity_dist?: SeverityDist;     // NG augmentation severity weights
  ok_synth_target?: number;         // Top-up each char to N OK via font-render synth (0=off)
  centroid_temperature?: number;    // Sigmoid temperature for centroid algo (default 5.0)
  include_imported_chars?: boolean; // Merge labeled crops from Imported Chars pool (default true)
}

export interface MLModelMetrics {
  accuracy_train: number;
  accuracy_test: number;
  n_ok: number;
  n_ng: number;
  n_total: number;
  confusion_matrix: number[][];
  report: string;
  golden_chars?: string[];   // chars with golden templates (v2 models only)
}

export interface MLModel {
  id: string;
  project_id: string;
  algorithm: string;
  params: Record<string, unknown>;
  augment_factor: number;
  metrics: MLModelMetrics;
  model_path: string;
  status: 'pending' | 'training' | 'completed' | 'failed';
  error?: string;
  created_at: string;
  phase?: string | null;
  progress?: number;
}

export interface TrainingLogEntry {
  idx: number;
  ts: number;          // unix epoch seconds (server clock)
  level: string;       // "INFO" | "WARNING" | "ERROR" ...
  msg: string;
}

export interface TrainingLogResponse {
  logs: TrainingLogEntry[];
  next_since: number;
  phase: string | null;
  progress: number;
  status: 'pending' | 'training' | 'completed' | 'failed';
  error: string | null;
}

export interface PredictResult {
  id: string;
  x: number; y: number; w: number; h: number;
  prob_ok: number;
  label: 'OK' | 'NG';
  crop_b64: string;
  char_id?: string | null;
  aligned_b64?: string | null;   // input after alignment to golden (v2 + golden)
  golden_b64?: string | null;    // reference golden for this char
  diff_b64?: string | null;      // JET-colormap diff heatmap
}

export interface CharCoverageResponse {
  covered: string[];
  missing: string[];
  coverage_pct: number;
  model_chars: string[];
}

export interface TestSetCropResult {
  crop_b64: string;
  true_label: 'OK' | 'NG';
  pred_label: 'OK' | 'NG';
  prob_ok: number;
  correct: boolean;
  char_id?: string | null;
}

// ──────── Static URL builders ──────────────────────────────────────────────

/** Full URL to a camera snapshot image (served from /public/images_temp) */
export const cameraImageUrl = (filename: string) =>
  `${SERVER_ORIGIN}/api/camera-images/${encodeURIComponent(filename)}`;

/** Full URL to a project image (served from /public/ml_projects/{id}/images) */
export const projectImageUrl = (projectId: string, filename: string) =>
  `${SERVER_ORIGIN}/api/ml-files/${projectId}/images/${encodeURIComponent(filename)}`;

// ──────── API client ───────────────────────────────────────────────────────

export const mlTrainingAPI = {
  // Projects
  createProject: (name: string, description?: string) =>
    api.post<MLProject>('/ml/projects', { name, description }).then(r => r.data),

  listProjects: () =>
    api.get<MLProject[]>('/ml/projects').then(r => r.data),

  getProject: (id: string) =>
    api.get<MLProject>(`/ml/projects/${id}`).then(r => r.data),

  updateProject: (id: string, name: string, description?: string) =>
    api.patch<MLProject>(`/ml/projects/${id}`, { name, description }).then(r => r.data),

  deleteProject: (id: string) =>
    api.delete(`/ml/projects/${id}`).then(r => r.data),

  cloneProject: (id: string, name?: string, description?: string) =>
    api.post<MLProject>(`/ml/projects/${id}/clone`, { name, description }).then(r => r.data),

  // Snapshot camera buffer → stable temp folder for current session
  snapshotImages: () =>
    api.post<{ copied: number; filenames: string[] }>('/ml/snapshot-images').then(r => r.data),

  // Available images (from /public/images_temp snapshot)
  listAvailableImages: () =>
    api.get<{ filename: string }[]>('/ml/available-images').then(r =>
      r.data.map(img => ({ filename: img.filename, url: cameraImageUrl(img.filename) }))
    ),

  // Project images
  listProjectImages: (projectId: string) =>
    api.get<{ filename: string; has_annotation: boolean }[]>(`/ml/projects/${projectId}/images`).then(r =>
      r.data.map(img => ({ ...img, url: projectImageUrl(projectId, img.filename) }))
    ),

  copyImages: (projectId: string, filenames: string[]) =>
    api.post(`/ml/projects/${projectId}/images/copy`, { filenames }).then(r => r.data),

  uploadImages: (projectId: string, files: File[]) => {
    const form = new FormData();
    files.forEach(f => form.append('files', f));
    return api.post(`/ml/projects/${projectId}/images/upload`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then(r => r.data);
  },

  deleteImage: (projectId: string, filename: string) =>
    api.delete(`/ml/projects/${projectId}/images/${encodeURIComponent(filename)}`).then(r => r.data),

  getImageMeta: (projectId: string, filename: string) =>
    api.get<{ filename: string; width: number; height: number }>(
      `/ml/projects/${projectId}/images/${encodeURIComponent(filename)}/meta`
    ).then(r => ({ ...r.data, url: projectImageUrl(projectId, filename) })),

  // Segmentation
  segmentRegion: (projectId: string, filename: string, region: { x: number; y: number; w: number; h: number }) =>
    api.post<{ segments: CharSegment[]; count: number }>(
      `/ml/projects/${projectId}/segment`,
      { filename, region }
    ).then(r => r.data),

  // Annotations
  getAnnotation: (projectId: string, filename: string) =>
    api.get<AnnotationData>(
      `/ml/projects/${projectId}/annotations/${encodeURIComponent(filename)}`
    ).then(r => r.data),

  saveAnnotation: (projectId: string, filename: string, regions: AnnotationRegion[]) =>
    api.put<AnnotationData>(
      `/ml/projects/${projectId}/annotations/${encodeURIComponent(filename)}`,
      { regions }
    ).then(r => r.data),

  // Labeled crops (Train tab)
  getLabeledCrops: (projectId: string) =>
    api.get<{ crops: LabeledCrop[]; count: number }>(
      `/ml/projects/${projectId}/labeled-crops`
    ).then(r => r.data),

  // Preview synthetic NG (generated from OK samples). OK augmentation removed.
  previewSynthetic: (projectId: string, augmentFactor: number,
                     label: 'NG' = 'NG', severityDist?: SeverityDist) =>
    api.post<{ crops: SyntheticCrop[]; count: number }>(
      `/ml/projects/${projectId}/preview-synthetic`,
      { augment_factor: augmentFactor, label, severity_dist: severityDist }
    ).then(r => r.data),

  // Preview synthetic OK (font-render → composite on real BG → camera noise)
  previewSyntheticOk: (projectId: string, opts: {
    target_n_per_char?: number;
    only_below_threshold?: boolean;
    char_filter?: string[] | null;
    rotation_max_deg?: number;
    size_jitter?: number;
  }) =>
    api.post<{ crops: SyntheticOkCrop[]; count: number }>(
      `/ml/projects/${projectId}/preview-synthetic-ok`, opts
    ).then(r => r.data),

  // Training
  startTraining: (projectId: string, req: TrainRequest) =>
    api.post<{ model_id: string; status: string }>(
      `/ml/projects/${projectId}/train`, req
    ).then(r => r.data),

  listModels: (projectId: string) =>
    api.get<MLModel[]>(`/ml/projects/${projectId}/models`).then(r => r.data),

  getModelStatus: (projectId: string, modelId: string) =>
    api.get<MLModel>(`/ml/projects/${projectId}/models/${modelId}/status`).then(r => r.data),

  getTrainingLogs: (projectId: string, modelId: string, since: number = 0) =>
    api.get<TrainingLogResponse>(
      `/ml/projects/${projectId}/models/${modelId}/logs`,
      { params: { since } },
    ).then(r => r.data),

  // Prediction
  predict: (projectId: string, file: File, modelId?: string) => {
    const form = new FormData();
    form.append('file', file);
    if (modelId) form.append('model_id', modelId);
    return api.post<{ model_id: string; algorithm: string; results: PredictResult[] }>(
      `/ml/projects/${projectId}/predict`, form,
      { headers: { 'Content-Type': 'multipart/form-data' } }
    ).then(r => r.data);
  },

  getTestSetCrops: (projectId: string, modelId: string) =>
    api.get<{ crops: TestSetCropResult[]; count: number }>(
      `/ml/projects/${projectId}/models/${modelId}/test-set-crops`
    ).then(r => r.data),

  // Check which required chars are covered by the model
  charCoverage: (projectId: string, modelId: string, chars: string[]) =>
    api.get<CharCoverageResponse>(
      `/ml/projects/${projectId}/models/${modelId}/char-coverage`,
      { params: { chars: chars.join(',') } },
    ).then(r => r.data),

  // Full training report (metrics + test-set with base64 images)
  getModelReport: (projectId: string, modelId: string, opts?: {
    include_testset?: boolean;
  }) =>
    api.get(
      `/ml/projects/${projectId}/models/${modelId}/report`,
      { params: opts },
    ).then(r => r.data),

  // ── Active learning: search candidate chars from inspection history ──
  inspectionCandidates: (projectId: string, opts: {
    recipe_id?: string;
    date_from?: string;
    date_to?: string;
    include_pred_ok?: boolean;
    include_pred_ng?: boolean;
    limit?: number;
  }) =>
    api.get<{ candidates: InspectionCandidate[]; count: number }>(
      `/ml/projects/${projectId}/inspection-candidates`,
      { params: opts },
    ).then(r => r.data),

  // ── Imported Chars pool ──
  listCharImportBatches: (projectId: string) =>
    api.get<CharImportBatch[]>(`/ml/projects/${projectId}/char-imports/batches`).then(r => r.data),

  createCharImportBatch: (
    projectId: string,
    selections: Array<{ inspection_id: string; annotation_idx: number }>,
    batch_name?: string,
  ) =>
    api.post<{
      batch_id: string | null;
      batch_name?: string;
      imported: number;
      skipped: number;
      errors: Array<{ inspection_id: string; reason: string }>;
    }>(`/ml/projects/${projectId}/char-imports/batches`, { selections, batch_name }).then(r => r.data),

  renameCharImportBatch: (projectId: string, batchId: string, name: string) =>
    api.patch<{ id: string; name: string }>(
      `/ml/projects/${projectId}/char-imports/batches/${batchId}`, { name }
    ).then(r => r.data),

  deleteCharImportBatch: (projectId: string, batchId: string) =>
    api.delete(`/ml/projects/${projectId}/char-imports/batches/${batchId}`).then(r => r.data),

  listCharImports: (projectId: string, opts?: { batch_id?: string; label?: 'OK' | 'NG' }) =>
    api.get<CharImportItem[]>(`/ml/projects/${projectId}/char-imports/chars`, {
      params: opts,
    }).then(r => r.data),

  updateCharImport: (projectId: string, charId: string, update: { char_id?: string | null; label?: 'OK' | 'NG' }) =>
    api.patch<CharImportItem>(
      `/ml/projects/${projectId}/char-imports/chars/${charId}`, update,
    ).then(r => r.data),

  deleteCharImport: (projectId: string, charId: string) =>
    api.delete(`/ml/projects/${projectId}/char-imports/chars/${charId}`).then(r => r.data),

  bulkCharImports: (
    projectId: string,
    body: { char_ids: string[]; label?: 'OK' | 'NG'; delete?: boolean },
  ) =>
    api.patch<{ updated?: number; deleted?: number }>(
      `/ml/projects/${projectId}/char-imports/chars/bulk`, body,
    ).then(r => r.data),
};

export interface InspectionCandidate {
  inspection_id: string;
  recipe_id: string;
  recipe_name: string;
  camera_serial: string;
  frame_idx: number;
  annotation_idx: number;
  expected: string;
  ml_label: 'OK' | 'NG';
  ml_p_ok: number;
  timestamp: string | null;
  image_path: string;
  crop_b64: string;
  // Set when this (inspection_id, annotation_idx) pair is already in the
  // project's char-import pool — FE shows a "Already imported in <batch>"
  // badge and disables the checkbox.
  imported_batch_id: string | null;
  imported_batch_name: string | null;
}

export interface CharImportBatch {
  id: string;
  name: string;
  created_at: string;
  total: number;
  ok_count: number;
  ng_count: number;
}

export interface CharImportItem {
  id: string;
  batch_id: string;
  char_id: string | null;
  label: 'OK' | 'NG';
  crop_url: string;
  ml_label: 'OK' | 'NG';
  ml_p_ok: number;
  inspection_id: string;
  annotation_idx: number;
  recipe_name: string | null;
  camera_serial: string;
  frame_idx: number;
  source_timestamp: string | null;
  created_at: string;
}
